src/attr/_make.py

```diff
@@ -2120,7 +2153,7 @@ def fmt_setter_with_converter(
         )
         arg_name = a.name.lstrip("_")
 
-        has_factory = isinstance(a.default, Factory)
+        has_factory = Factory is not None and isinstance(a.default, Factory)
```

Comment:
I don't understand this and why you set `Factory` globally to `None`. Could you explain please?

Reply:
`Factory` is an `attrs` class, so it calls this code while it is being declared. If we don't forward-declare `Factory`, this will lead to some sort of undeclared variable error (I forget exactly what the error text was). 

However, we must obviously abide by the constraint of not using `factory=` inside the declaration of `Factory`, so that `has_factory` could never be True while `Factory` was being constructed; by the time a user class is being constructed, `Factory is not None`, so the code falls back to checking `isinstance(a.default, Factory)`.

Open to other suggestions here.

Reply:
Ah shit.

How about we pull this on Factory:

https://github.com/indigoviolet/attrs/blob/0effbc2852c6d22e6d351d10b56ed5bc12d001d8/src/attr/_make.py#L2546-L2568

The biggest downside of working on attrs is not being able to use attrs. 🤪

Reply:
Hmm- I'm still a bit confused by that - does Factory need to be frozen like `Attribute`? Does it need a get/set state, a cmp? Not sure I understand the consequences here completely.

Reply:
I'm not saying you should copy paste. What I mean is that I'd like you to re-create Factory (as it is right now) using the `_add_XXX` primities like `Attribute`. Sorry for being fuzzy.

Reply:
Yes, I understood that, but I'm not exactly sure which features used in the Attribute class are needed to correctly create the Factory class. For example, does it need to be frozen, does it need to support pickling etc.

Reply:
I would say go for the same properties as now:

- slots
- hash
- eq
- repr

I think we can drop cmp; that makes no sense.

It could be frozen but that would probably break someone so please don't.

Reply:
Maybe we should also add a test that its pickleable, if it doesn't already exist.

Reply:
ok, implemented your suggestion; lmk if I missed something. I think there's already a [test for pickling](https://github.com/python-attrs/attrs/blob/master/tests/test_functional.py#L272-L294).

Reply:
Looks like mypy tests are failing on master, see https://github.com/python-attrs/attrs/pull/748